### PR TITLE
Add stubs for whatthepatch

### PR DESCRIPTION
**Pull request by cuspymd**
Hi, I added stubs for [whatthepatch](https://github.com/cscorley/whatthepatch).

### CODE REVIEW

File: stubs/whatthepatch/whatthepatch/exceptions.pyi [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from typing import Any
+
+class WhatThePatchException(Exception): ...
+
+class HunkException(WhatThePatchException):
+    hunk: Any
```
Comment (Akuli) [inline stubs/whatthepatch/whatthepatch/exceptions.pyi:6]:
```suggestion
    hunk: int | None
```

File: stubs/whatthepatch/whatthepatch/patch.pyi [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    line: int | None
+    hunk: int
+
+file_timestamp_str: str
+diffcmd_header: Any
+unified_header_index: Any
+unified_header_old_line: Any
+unified_header_new_line: Any
+unified_hunk_start: Any
+unified_change: Any
```
Comment (Akuli) [inline stubs/whatthepatch/whatthepatch/patch.pyi:31]: These could be annotated as `Pattern[str]`, where `Pattern` is imported from `typing`.

File: stubs/whatthepatch/whatthepatch/patch.pyi [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+cvs_header_index: Any
+cvs_header_rcs: Any
+cvs_header_timestamp: Any
+cvs_header_timestamp_colon: Any
+old_cvs_diffcmd_header: Any
+
+def parse_patch(text: str) -> Iterator[diffobj]: ...
+def parse_header(text: str) -> header | None: ...
+def parse_scm_header(text: str) -> header | None: ...
+def parse_diff_header(text: str) -> header | None: ...
```
Comment (Akuli) [inline stubs/whatthepatch/whatthepatch/patch.pyi:65]: All of these functions seem to accept `str | Iterable[str]` too.

File: stubs/whatthepatch/whatthepatch/snippets.pyi [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from collections.abc import Iterable
+from typing import Any
+
+def remove(path: str) -> None: ...
```
Comment (Akuli) [inline stubs/whatthepatch/whatthepatch/snippets.pyi:4]: This function actually works with any `StrOrBytesPath` (defined in `_typeshed`). But this doesn't matter much if the rest of the library only works with string paths, not e.g. bytes paths or `pathlib.Path` objects.

Comment (cuspymd) [reply stubs/whatthepatch/whatthepatch/snippets.pyi:4]: Thanks for your detailed code reviews.

**Review by Akuli — APPROVED** [review]
Thanks! I fixed the CI failures.

**Pull request merged**